Auditing projects for package vulnerabilities during restore

JonDouglas opened the pull request:
Introduction to an auditing experience at package restore time where known vulnerabilities will be reported.

This proposal introduces new MSBuild warnings and output to the NuGet restore operation.

[Rendered Spec](https://github.com/NuGet/Home/blob/dev-feature-vinr/proposed/2022/vulnerabilities-in-restore.md)

Please 👍 or 👎 this comment to help us with the direction of this feature & leave as much feedback/questions/concerns as you'd like on this issue itself and we will get back to you shortly.

Thank You 🎉

nkolev92 committed:
Add the initial version

JonDouglas committed:
Clean up and add prior art / futures

JonDouglas committed:
Update metadata

nkolev92 committed:
remove start date

nkolev92 committed:
Change should to may

JonDouglas committed:
Update vulnerabilities-in-restore.md

mattzink commented:
We use Nexus repo manager and Artifactory as local Nuget.org package mirrors, so it would be great if there was a way in the nuget.config file to specify that Nuget.org is used only for vulnerability lookup, not for package hosting.

loic-sharma commented on proposed/2022/vulnerabilities-in-restore.md:
````diff
@@ -0,0 +1,391 @@
+In cases where no known vulnerabilities are found, restore will provide a normal verbosity message for every project:
+
+```bash
+No known vulnerabilties found for <project path>.
````
```suggestion
No known vulnerabilities found for <project path>.

```

loic-sharma commented on proposed/2022/vulnerabilities-in-restore.md:
````diff
@@ -0,0 +1,391 @@
+| NU1904 | critical |
+
+```bash
+Found 2 vulnerabilities (0 low, 1 moderate, 0 high, 1 critical) in 2 package(s)
````
What's the value in reporting severities with 0 hits? Would it be better to exclude those instead?

```suggestion
Found 2 vulnerabilities (1 moderate, 1 critical) in 2 package(s)

```

JonDouglas replied:
Good idea, let's keep this open for discussion!

loic-sharma commented on proposed/2022/vulnerabilities-in-restore.md:
```diff
@@ -0,0 +1,391 @@
+  - `@name`, a user friendly name for the page, **must** be unique.
+  - `@id`, the url that contains the data.
+  - `@updated`, a UTC timestamp when the content at `@id` was updated last.
+  - `comment`, a user friendly description.
```
What's the value for this field? Does this really need to be required? Will this be noisy repetitive values?

loic-sharma commented on proposed/2022/vulnerabilities-in-restore.md:
```diff
@@ -0,0 +1,391 @@
+  - The second page, `update`, represents the data from the last update of `base`, to present. This page **should** be updated with new data.
+  - Periodically, for example once a month, the data from `update` **should** be migrated to `base`.
+  - If an entry needs to be removed from `base`, `base` should be updated.
+- Other partioning strategies are workable as well, as long as only 2 pages are receiving all the new data.  
```
Are the `base` and `update` names required if using other partitioning strategies?

Also, what's the value in requiring a maximum of two pages? Does it complicate the client logic to support arbitrary number of pages?

nkolev92 replied:
Base and update aren't required names, but highly recommended.

The idea with 2 pages updating is that the client will always want to have all the data be up to date. 
It would allow refreshing to be super fast without an overly complicated protocol, as we'd only download the out of date files.

It's balance between CI (where the data is always download) and local, where the data is refreshed.

loic-sharma replied:
The 2 page idea makes perfect sense. I'm more curious as to why we *require* a maximum of 2 pages; that seems odd.

nkolev92 replied:
It's not really suggesting maximum of 2 pages.
Or at least, it's not supposed to :) 

What that line is trying to communicate is that as few pages as possible should be updated, so that the client is not forced to redownload pages all the time.

zivkan replied:
I updated the spec to try to make it clearer that this is our suggestion, but servers can implement the data partitioning in different ways. The client will just look at the list of files, and their update timestamps to determine whether it can use a previously cached download or not.

JonDouglas committed:
Update proposed/2022/vulnerabilities-in-restore.md

Co-authored-by: Loïc Sharma <737941+loic-sharma@users.noreply.github.com>

loic-sharma commented on proposed/2022/vulnerabilities-in-restore.md:
```diff
@@ -0,0 +1,391 @@
+- The vulnerability resource **must** at minimum contain 1 page.
+- The pages within the vulnerability resource **must** be exclusive.
+- The vulnerability resource may 2 pages.
+  - One page, `base`, represents the data up to a certain point in time. This page **must** be updated periodically, based on a trigger.
```
How big will nuget.org's `base` page be? How quickly will it grow? In the worst case scenario (like an Australian customer), are we concerned about the performance impact of updating this `base` page during a restore?

nkolev92 replied:
These were the numbers a few weeks ago:

vulnerability_all_alternate.json    2.4 Mb
vulnerability_all_alternate.json.gz 78.9 Kb

> are we concerned about the performance impact of updating this base page during a restore?

Definitely a consideration. We'd only redownload when restore actually happens. The whole process takes time, but right now we think those numbers are relatively small and a cost worth paying given the benefit.

zivkan replied:
FWIW, I tested a different file format for the vulnerability data, trading off computation speed vs file size, with the idea that smaller file size will parse more quickly, so the slower per-package computation might still come ahead. Benchmarks suggest yes, overall perf is better, and the current file size for all github reviewed vulnerabilities is 175KB uncompressed, 9KB gzipped. So, we'll be going ahead with the smaller file size format instead now.

loic-sharma commented on proposed/2022/vulnerabilities-in-restore.md:
```diff
@@ -0,0 +1,391 @@
+
+### Allow sources to upstream vulnerability information
+
+While the NuGet tooling does support multiple sources and a means to control which packages come from which source through [Package Source Mapping](https://learn.microsoft.com/en-us/nuget/consume-packages/package-source-mapping), all configurations are not going to contain nuget.org, and thus won't easily get the data curated from the [GitHub Advisory Database](https://devblogs.microsoft.com/nuge ...
```
Will vulnerability reporting take source maps into account? If two package sources report different vulnerabilities for the same package, will the client report all vulnerabilities?

nkolev92 replied:
That's a great question. 

Right now, we think no is the answer for that, we "merge" the data, but we can certainly change that based on community input.

loic-sharma commented on proposed/2022/vulnerabilities-in-restore.md:
```diff
@@ -0,0 +1,391 @@
+  - Experimentation is arguably easier as we're not introducing new syntax, at least not immediately.
+  - The challenge is that we'd need to establish a protocol, and something that could be slower on the uptake.
+
+### Optimizing vulnerability check cadence
```
Would it make sense to update vulnerability info in the background in Visual Studio? That way, restore performance in Visual Studio wouldn't be affected by this feature.

kapsiR replied:
Good idea!  
Additionally this could be another step in CI (cli), so that not every build would be affected by a new download or in other words: 
Have more control over the time when a new file is going to be downloaded

nkolev92 replied:
> Would it make sense to update vulnerability info in the background in Visual Studio? That way, restore performance in Visual Studio wouldn't be affected by this feature.

We don't expect this to be taking so long where it's noticeable consistently. 

On the topic of background downloading, it's not always desirable to communicate with sources without user intent. That could for example trigger authentication and we wouldn't want anything to interrupt the user. 

> Additionally this could be another step in CI (cli), so that not every build would be affected by a new download or in other words:

There are a lot of optimizations that happen prior to vulnerability data being downloaded. 
If you're not changing your packages/deleting the obj folder, restore will be snappy and won't redownload sources.

loic-sharma commented on proposed/2022/vulnerabilities-in-restore.md:
```diff
@@ -0,0 +1,391 @@
+### Providing vulnerability information through a json, remotely or locally
+
+Other components may want to provide vulnerability information that's not tied to a source.
+Whatever approach we take there will be a way to represent the vulnerability information through a json file.
```
Perhaps there could be a nuget.config for a list of vulnerability info URLs? That'd be useful if I have a single private package source that doesn't point to its upstream vulnerability info URL yet.

nkolev92 replied:
Yes, it would have to be something along those lines.

kapsiR commented on proposed/2022/vulnerabilities-in-restore.md:
```diff
@@ -0,0 +1,392 @@
+
+## Summary
+
+Audit your dependencies when restoring your NuGet packages with security vulnerabilities reported to GitHub Advisory Database.
```
I don't know exactly what sources this database is using but are there any plans to support additional databases too? (e.g. Google osv.dev)

zivkan replied:
The current proposal is that NuGet client gets the list of vulnerabilities from the NuGet package source. How the server implements the vulnerability information resource is up to the server implementation. They certainly could use osv.dev.

Looking at osv.dev, I don't see a way to download the entire database (unless I need to clone, build, and run the github repo?). So, from a NuGet client point of view, osv.dev doesn't fit well with our design.

This could be something added in the future if there's enough demand, but the current plan is that NuGet server implementations are responsible for providing the client with the relevant data.

kapsiR replied:
Thanks to clarify on that, maybe it's worth adding to the docs? 

FYI there are data dumps available:
https://github.com/google/osv.dev#data-dumps

kapsiR replied:
And more important: GH advisory db is a source of osv.dev! 😉

zivkan replied:
I updated the spec to try to make this clearer.

JonDouglas committed:
Update vulnerabilities-in-restore.md

CamiloTerevinto commented:
This sounds very interesting, thanks @JonDouglas. I have a few comments:

* CLI switches to run/skip auditing 

It's quite common for an entire build to be run twice (feature branch and dev/main branch) for a change in environments with CI/CD. I think that for this to be useful, we should have an easy way of telling `dotnet restore` to run or skip the audit (for example, run the audit on the PR build, skip it in the dev build).

* Fail on X level

It would be useful to have a way to make `dotnet restore` exit as failure when a package with a particular vulnerability level is found (such as treating `NU1904` as error instead of warning).  
However, this might be doable already through something like `<WarningsAsErrors>NU1904</WarningsAsErrors>`.

nkolev92 commented:
> CLI switches to run/skip auditing

Any CLI switch basically doubles as an MSBuild property. We're proposing MSBuild properties for opting into auditing, so that automatically makes it available as a CLI switch. Example: `dotnet restore -p:NuGetAudit=disable`

zivkan committed:
Update proposed file format

zivkan committed:
Be more explicit about data source

nuget.org uses GitHub's advisories database, but other servers can use
other data sources if they want

zivkan committed:
Move command line output to future possibilities

zivkan committed:
minor update

zivkan committed:
try to make protocol proposal a little more clear

zivkan commented on proposed/2022/vulnerabilities-in-restore.md:
```diff
@@ -0,0 +1,406 @@
+
+It is **very important* that if the upstream sources link to the nuget.org vulnerability, the actual nuget.org urls are used wherever possible, as this would allow the client to easily deduplicate the vulnerability information.
+
+### Command Line Restore output
```
@JonDouglas FYI, I moved the vulnerability summary output on command line restores under "Future Possibilities" for two reasons. 1. So we can get a V1 of this feature out sooner. 2. I don't believe this proposal has been specced out well. I have a bunch of questions if we want to implement it.

For example, in the sample output, if there are non-vulerability warnings or errors, where in the output do they go?  Before the vulnerability summary, or below with the sample vulerability warnings?   If before, should vulerability warnings/errors be output twice, once between the summary once after?

What's the expected order of output when restoring a solution with many projects? Wait until all projects finish restoring, and then output all their summaries?  Or output project summaries as each project finishes restoring, even if it means that the vulnerability summary for that project will be somewhere much earlier in the restore log?

When restoring a solution where the same package version is in multiple projects, should we de-duplicate the output?

Anyway, as I said, I think we can take this on in the future if we still think it's useful. Let's get a V1 of the feature done without it, and see how we and customers feel about it.

nkolev92 replied:
I think some details, like the ones about the handling of multiple vulnerability for a package can still be in the main part as that's likely going to be done.

JonDouglas replied:
Let's see what this shapes to be while working to answer the unknown. Thanks for being so thorough with questions to be answered.

nkolev92 commented on proposed/2022/vulnerabilities-in-restore.md:
````diff
@@ -0,0 +1,406 @@
+
+When there are packages used by the project with vulnerabilities, they will be outjust like like any other warning or error:
+
+```text
````
If you put bash, you get better highlighting :) Not a big deal :D

zivkan replied:
The only highlighting I see are the words `in`, `for` and the `<` and `>` characters. Which make sense in bash scripts, but in my opinion it doesn't make sense for these samples.  Why in the sentence "Found 2 vulerabilities in 1 page(s)" should only the word `in`, be highlighted?  🤷

nkolev92 replied:
Here's what I am seeing: 

![image](https://user-images.githubusercontent.com/2878341/218845920-fde7720d-77f8-4427-a761-62c7a83c82ff.png)

The quoted part is being highlighted. It's not a big deal either way, and most definitely a nit :) Your call.

nkolev92 commented on proposed/2022/vulnerabilities-in-restore.md:
```diff
@@ -0,0 +1,406 @@
+- The objects within the array of objects **must** contain:
+  - Severity, an `int`, where `0` means `low`, `1` means `medium`, `2` means `high`, `3` means `critical`
+  - Advisory url, a url.
+  - Versions, a version range in [NuGet range syntax](https://learn.microsoft.com/nuget/concepts/package-versioning#version-ranges) of affected package versions.
```
maybe mention that it includes the prerelease versions in the range as well?

nkolev92 commented on proposed/2022/vulnerabilities-in-restore.md:
````diff
@@ -0,0 +1,406 @@
+warning NU1902: Package 'Contoso.Forms' 8.4.1 has a known moderate severity vulnerability, https://github.com/advisories/GHSA-1234-5678-9012.
+```
+
+If a package has more than 1 vulnerability, each vulnerability will have its own warning. These security issues are often likely independent and may require independent fixes.
````
This can probably be in the main part.

nkolev92 reviewed (commented):
Great updates @zivkan

zivkan committed:
Hopefully these changes are improvements.

JonDouglas commented:
We have approval from @nkolev92, but @zivkan did you want to approve before merge?

zivkan commented on proposed/2022/vulnerabilities-in-restore.md:
```diff
@@ -0,0 +1,421 @@
+
+  {
+    "@id": "https://api.nuget.org/v3/vulnerability.json",
+    "@type": "VulnerabilityInfo/6.3.0",
```
```suggestion
    "@type": "VulnerabilityInfo/6.6.0",
```

Oops, I already changed this elsewhere in the doc, but I missed it here.

zivkan committed:
fix service index resource client version
